compiler/rustc_const_eval/src/const_eval/valtrees.rs

```diff
@@ -0,0 +1,479 @@
+                        ty::ValTree::Leaf(_) => {
+                            let mut place = create_mplace_from_layout(
+                                &mut ecx,
+                                ty::ParamEnv::empty().and(*inner_ty),
+                            );
+
+                            fill_place_recursively(&mut ecx, &mut place, valtree, *inner_ty);
+                            dump_place(&ecx, place.into());
+
+                            let ref_place = place.mplace.to_ref(&tcx);
+                            let imm = ImmTy::from_immediate(
+                                ref_place,
+                                tcx.layout_of(param_env_ty).unwrap(),
+                            );
+
+                            op_to_const(&ecx, &imm.into())
+                        }
+                    }
+                }
+            }
+        }
+        ty::Tuple(_) | ty::Array(_, _) | ty::Adt(..) => {
+            let mut place = create_mplace_from_layout(&mut ecx, param_env_ty);
+            debug!(?place);
+
+            fill_place_recursively(&mut ecx, &mut place, valtree, ty);
+            dump_place(&ecx, place.into());
+            intern_const_alloc_recursive(&mut ecx, InternKind::Constant, &place).unwrap();
+
+            let const_val = op_to_const(&ecx, &place.into());
+            debug!(?const_val);
+
+            const_val
+        }
+        ty::Never
+        | ty::FnDef(..)
+        | ty::Error(_)
+        | ty::Foreign(..)
+        | ty::Infer(ty::FreshIntTy(_))
+        | ty::Infer(ty::FreshFloatTy(_))
+        | ty::Projection(..)
+        | ty::Param(_)
+        | ty::Bound(..)
+        | ty::Placeholder(..)
+        | ty::Opaque(..)
+        | ty::Infer(_)
+        | ty::Closure(..)
+        | ty::Generator(..)
+        | ty::GeneratorWitness(..)
+        | ty::FnPtr(_)
+        | ty::RawPtr(_)
+        | ty::Str
+        | ty::Slice(_)
+        | ty::Dynamic(..) => bug!("no ValTree should have been created for type {:?}", ty.kind()),
+    }
+}
+
+// FIXME Needs a better/correct name
+#[instrument(skip(ecx), level = "debug")]
+fn fill_place_recursively<'tcx>(
+    ecx: &mut CompileTimeEvalContext<'tcx, 'tcx>,
+    place: &mut MPlaceTy<'tcx>,
+    valtree: ty::ValTree<'tcx>,
+    ty: Ty<'tcx>,
+) {
+    // This will match on valtree and write the value(s) corresponding to the ValTree
+    // inside the place recursively.
+
+    let tcx = ecx.tcx.tcx;
+
+    match ty.kind() {
+        ty::Bool | ty::Int(_) | ty::Uint(_) | ty::Float(_) | ty::Char => {
+            let scalar_int = valtree.unwrap_leaf();
+            debug!("writing trivial valtree {:?} to place {:?}", scalar_int, place);
+            ecx.write_immediate(
+                Immediate::Scalar(ScalarMaybeUninit::Scalar(scalar_int.into())),
+                &(*place).into(),
+            )
+            .unwrap();
+        }
+        ty::Ref(_, inner_ty, _) => {
+            match inner_ty.kind() {
+                ty::Slice(_) | ty::Str => {
```

Comment:
The fallback arm should mostly do the right thing for slices, too. So you should be able to deduplicate them and just add a small adjustment for the metadata within the logic